spec/system/admin/products_v3/products_spec.rb

```diff
@@ -18,9 +18,128 @@
   let(:categories_search_selector) { 'input[placeholder="Search for categories"]' }
   let(:tax_categories_search_selector) { 'input[placeholder="Search for tax categories"]' }
 
-  it "can see the new product page" do
-    visit admin_products_url
-    expect(page).to have_content "Bulk Edit Products"
+  describe "with no products" do
+    before { visit admin_products_url }
+    it "can see the new product page" do
+      expect(page).to have_content "Bulk Edit Products"
+      expect(page).to have_text "No products found"
+      # displays buttons to add products with the correct links
+      expect(page).to have_link(class: "button", text: "New Product", href: "/admin/products/new")
+      expect(page).to have_link(class: "button", text: "Import multiple products",
+                                href: "/admin/products/import")
+    end
+  end
+
+  describe "using the page" do
+    describe "using column display dropdown" do
+      let(:product) { create(:simple_product) }
+
+      before do
+        pending "Pending implementation, issue #11055"
+        login_as_admin
+        visit spree.admin_products_path
+      end
+
+      it "shows a column display dropdown, which shows a list of columns when clicked" do
+        expect(page).to have_selector "th", text: "NAME"
+        expect(page).to have_selector "th", text: "PRODUCER"
+        expect(page).to have_selector "th", text: "PRICE"
+        expect(page).to have_selector "th", text: "ON HAND"
+
+        toggle_columns /^.{0,1}Producer$/i
+
+        expect(page).not_to have_selector "th", text: "PRODUCER"
+        expect(page).to have_selector "th", text: "NAME"
+        expect(page).to have_selector "th", text: "PRICE"
+        expect(page).to have_selector "th", text: "ON HAND"
+      end
+    end
+  end
+
+  describe "listing" do
+    let!(:p1) { create(:product) }
+    let!(:p2) { create(:product) }
+
+    before do
+      visit admin_products_url
+    end
+
+    it "displays a list of products" do
+      within ".products" do
+        # displays table header
+        expect(page).to have_selector "th", text: "Name"
+        expect(page).to have_selector "th", text: "SKU"
+        expect(page).to have_selector "th", text: "Unit scale"
+        expect(page).to have_selector "th", text: "Unit"
+        expect(page).to have_selector "th", text: "Price"
+        expect(page).to have_selector "th", text: "On Hand"
+        expect(page).to have_selector "th", text: "Producer"
+        expect(page).to have_selector "th", text: "Category"
+        expect(page).to have_selector "th", text: "Tax Category"
+        expect(page).to have_selector "th", text: "Inherits Properties?"
+        expect(page).to have_selector "th", text: "Actions"
+
+        # displays product list
+        expect(page).to have_field("_products_0_name", with: p1.name.to_s)
+        expect(page).to have_field("_products_1_name", with: p2.name.to_s)
```

Comment:
For this purpose, I created the helper `row_containing_name`, which describes it from the user's point of view. Instead of referring to hidden IDs, it asks: is there a table row containing a "Name" field with value of "x"? 
I think we could use it like:

```suggestion
        expect(page).to have_selector row_containing_name(p1.name.to_s)
        expect(page).to have_selector row_containing_name(p2.name.to_s)
```

Reply:
Thanks for coming up with this nice helper :pray: If I understand correctly, the helper looks for the product, without associating it with a given row, right?

I think this is fine for this test case, but we might need to associate rows with products, if we wish to test for ordering by name, for example (I see now that this test is missing from both the legacy and the BUU spec :eyes: )

Reply:
> If I understand correctly, the helper looks for the product, without associating it with a given row, right?

It's actually just a CSS selector that looks for a table row with a "Name"
I've coded the specs so that they interact with the interface. The user interface doesn't have row labels to say "this row is a product", or "this row is a variant", but it does have column labels (eg "Name"). 
So this helper is actually just a CSS selector that looks for a table row that has the specified "Name". 

> I think this is fine for this test case, but we might need to associate rows with products, if we wish to test for ordering by name, for example.

We actually have another helper in the "sorting" section testing order of products already.
If a product and its variant are both named the same "Carrots", then the spec would have to consider the ordering to figure out which one is which (just like we humans have to). Of course, we could refer to the underlying code, but I try to avoid that where possible, because it helps ensure the test really succeeds for a human.